OpenRA.Game/Network/Nat.cs

```diff
@@ -0,0 +1,91 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Mono.Nat;
+
+namespace OpenRA.Network
+{
+	public enum NatStatus { Enabled, Disabled, NotSupported }
+
+	public class Nat
+	{
+		public NatStatus Status
+		{
+			get
+			{
+				return NatUtility.IsSearching ? natDevice != null ?
+					NatStatus.Enabled : NatStatus.NotSupported : NatStatus.Disabled;
+			}
+		}
+
+		Mapping mapping;
+		INatDevice natDevice;
+
+		public Nat()
+		{
+			if (Game.Settings.Server.DiscoverNatDevices)
+			{
+				NatUtility.DeviceFound += DeviceFound;
+				NatUtility.StartDiscovery();
+			}
+		}
+
+		readonly SemaphoreSlim locker = new SemaphoreSlim(1, 1);
+
+		async void DeviceFound(object sender, DeviceEventArgs args)
+		{
+			await locker.WaitAsync();
+			try
+			{
+				// Only interact with one at a time. Some support both UPnP and NAT-PMP.
+				natDevice = args.Device;
+
+				Log.Write("nat", "Device found: {0}", natDevice.NatProtocol);
+				Log.Write("nat", "Type: {0}", natDevice.GetType().Name);
+			}
+			finally
+			{
+				locker.Release();
+			}
+		}
+
+		public void ForwardPort(int listen, int external)
+		{
+			var lifetime = Game.Settings.Server.NatPortMappingLifetime;
```

Comment:
Does this mean that games that have not started and finished within 45 minutes will have their connection terminated? Or that servers that have been opened but not started within 45 minutes will no longer be joinable?

Surely we need some kind of keepalive to stop this from timing out for active games?

Reply:
Interestingly you can also set it to 0 which for UPnP means indefinite but for NAT-PMP defaults to 7,200 seconds. I set it to 10 hours now. So if the game crashes and someone decides to uninstall then the router will still cleanup after a day.